src/puppetlabs/puppetdb/http/command.clj

```diff
@@ -168,32 +167,38 @@
               (normalize-new-request req)
               (normalize-old-request req)))))
 
-(defn- realize-body [body max-command-size]
+(defn- stream-with-max-check
   "Returns the body as an in-memory string or byte-array, reading it
   if necessary.  Throws ::body-stream-overflow if the max-command-size
   is not false and not respected."
+  [body max-command-size]
   (if-not max-command-size
-    (if (instance? java.io.InputStream body)
-      (IOUtils/toByteArray body)
-      body)
     (cond
       (instance? java.io.InputStream body)
-      (IOUtils/toByteArray
-       (restrained-drained-stream body (long max-command-size)))
+      body
+
+      (string? body)
+      (java.io.ByteArrayInputStream. (.getBytes body "UTF-8"))
+
+      :else
+      (throw (Exception. (str "Unexpected body type: " (class body)))))
+    (cond
+      (instance? java.io.InputStream body)
+      (restrained-drained-stream body (long max-command-size))
```

Comment:
Good point(s), though I suspect we may have any number of other places we'll need to address if we want to be more careful on that front (e.g. if postgresql is on the same disk, then you could just feed it "too much" actual data?).